File: src/main/java/dev/psygamer/wireframe/event/api/IEventBus.java
package dev.psygamer.wireframe.event.api;

import dev.psygamer.wireframe.event.Event;

import java.lang.reflect.Method;
import java.util.function.Consumer;

public interface IEventBus {
	
	void register(final Object target);
	
	<T extends Event> void addListener(final Consumer<T> consumer);
	
	<T extends Event> void addListener(final Consumer<T> consumer, final Event.Priority priority);
	
	<T extends Event> void addListener(final Object instance, final Method method);
	
	<T extends Event> void addListener(final Object instance, final Method method, final Event.Priority priority);
	
	boolean post(final Event event);
}

File: src/main/kotlin/dev/psygamer/wireframe/event/api/IEventBus.kt
package dev.psygamer.wireframe.event.api

import java.lang.reflect.Method

interface IEventBus {

	fun register(target: Any)
	fun post(event: Event): Boolean

	fun <T : Event> addListener(consumer: (T) -> Unit, priority: EventPriority = EventPriority.NORMAL)
	fun addListener(instance: Any?, method: Method, priority: EventPriority = EventPriority.NORMAL)
}